File: graph/ruby/adjmat.rb
class Graph
    def initialize(nv, edges)
        @nvertices = nv
        @adjmat = nv.times.collect { Array.new(nv, 0) }
        edges.each do |i, j|
            @adjmat[i][j] = 1
        end
    end
    attr_reader :nvertices
    def each_vertices
        @nvertices.times do |i|
            yield i
        end
    end
    def each_edges
        @nvertices.times do |i|
            @nvertices.times do |j|
                yield [i,j] if @adjmat[i][j] == 1
            end
        end
    end
    def each_successors(i)
        @nvertices.times do |j|
            yield [j] if @adjmat[i][j] == 1
        end
    end
    def vertices
        Enumerator.new(self, :each_vertices)
    end
    def edges
        Enumerator.new(self, :each_edges)
    end
    def successors(i)
        Enumerator.new(self, :each_successors, i)
    end
end

File: graph/haskell/Adjmat.hs
module Adjmat (
    Graph, Vertex, Edge,
    create, nvertices, vertices, edges, successors
    ) where

import Data.Array

type Vertex = Int
type Edge   = (Vertex, Vertex)
type Graph  = Array Edge Int

create :: Int -> [Edge] -> Graph
create nv edges = accumArray (\_ _ -> 1) 0
    ((0,0), (nv-1,nv-1)) [(e, 1) | e <- edges]

nvertices :: Graph -> Int
nvertices g = n + 1
    where
    (_, (n, _)) = bounds g

vertices :: Graph -> [Vertex]
vertices g = [0..nvertices g]

edges :: Graph -> [Edge]
edges g = [ e | (e, 1) <- assocs g ]

successors :: Graph -> Vertex -> [Vertex]
successors g i = [ j | j <- [0..n], g!(i, j) == 1]
    where
    (_, (n, _)) = bounds g


